### ISSUE TITLE
Feature request: Plot pre-existing models with geom_smooth

### ISSUE THREAD

**Issue by tmalsburg** [tidyverse/ggplot2#4552]
**Background:** We can use `geom_smooth` to fit (e.g.) an `lm` to the data shown in a plot.  However, it's often useful to show models that were fit on a different data set.  For instance a model is fit on the training set but plotted on top of the evaluation set.  To my knowledge, there is currently no concise way to do this in ggplot2.

**Example:**  Currently, the default solution seems to be:

```r
library(tidyverse)
set.seed(1)

tibble(
  x = runif(20, 0, 10),
  y = 1 + 0.5 * x + rnorm(20, 0, 3)) -> d

lm(y~x, slice(d, 1:10)) -> m

slice(d, 11:20) %>%
  ggplot(aes(x,y)) +
  geom_point() +
  geom_abline(
    intercept = coef(m)[1],
    slope     = coef(m)[2])
```

<img src="https://user-images.githubusercontent.com/3817450/125060522-c17b1d80-e0ac-11eb-9c82-6e5a81f4e1b3.png" width="300">

**Proposal:** Allow users to pass a pre-existing model (`lm` or otherwise) to `geom_smooth` (e.g., `geom_smooth(model=m)`).  Then `geom_smooth` uses  the model's `predict` function to evaluate it for the x-range shown in the plot and displays the results.

You can probably think of a better solution but I provide the proposal above for clarification and as a starting point for discussion.

**Comment by yutannihilation** [comment]
Thanks for the suggestion, but I think this kind of feature is a better fit for an extension package.

Just in case you don't know these yet, you can use broom package to generate the fitted data,

``` r
library(tidyverse)
set.seed(1)

tibble(
  x = runif(20, 0, 10),
  y = 1 + 0.5 * x + rnorm(20, 0, 3)) -> d

lm(y~x, slice(d, 1:10)) -> m

library(broom)

d <- augment(m)

ggplot(d) +
  geom_point(aes(x, y)) +
  geom_line(aes(x, .fitted))
```

![](https://i.imgur.com/sdNheDH.png)

or you can also use ggfortify package to draw the nice plots magically.

``` r
library(ggfortify)

autoplot(m)
```

![](https://i.imgur.com/MDgxrax.png)

**Comment by tmalsburg** [comment]
Thanks for your response. But broom doesn't solve this problem satisfactorily.  Note that I'd like plot a model that was fit using one data set on top of a scatter plot of a different data set (training vs. test set scenario, very common in statistics and machine learning).  Your broom solution doesn't do that.  It shows the model and scatter plot for the *same* data set.  For that we already have `geom_smooth(method = lm)`, no extension package needed.

We can adapt your solution to show model and scatter plot for different data sets:

``` R
library(tidyverse)
library(broom)
set.seed(1)

tibble(
  x = runif(20, 0, 10),
  y = 1 + 0.5 * x + rnorm(20, 0, 3)) -> d

lm(y~x, slice(d, 1:10)) -> m

augment(m) -> d1

slice(d, 11:20) %>%
  ggplot(aes(x, y)) +
  geom_point() +
  geom_line(aes(x, .fitted), d1)
```

<img src="https://user-images.githubusercontent.com/3817450/125401618-add4fd00-e3b3-11eb-9d39-c5214d3cbe53.png" width="400">

However, there are two problems with that:
1. It's as much typing as the solution using `geom_abline` (see my original post above).
2. This solution may show a line that's too short (or not visible at all in extreme cases).  See the output above as an illustration: The line ends where the training data ends, not where the plot ends.

Of course, I could develop a new package that supports creating the desired plot, but wouldn't it be overkill to maintain a package for a common task that base-R can do via `plot(d); abline(m)`.

Background is that I'm teaching ggplot2 at my university and students complain that ggplot solutions are often much longer than base-R solutions.  `geom_smooth(model=m)` seems like a very natural and idiomatic solution for this specific case.

**Comment by tmalsburg** [comment]
I completely understand your point about adding complexity (I'm also maintaining software some software packages). But perhaps this is a case where the payoff / complexity ratio is reasonable enough to at least consider it more seriously.

**Comment by teunbrand** [comment]
I'm just going to leave a comment here that `stat_function()` + `predict()` can handle most of the `y ~ x` type of models.

``` r
library(tidyverse)
set.seed(1)

tibble(
  x = runif(20, 0, 10),
  y = 1 + 0.5 * x + rnorm(20, 0, 3)) -> d

lm(y~x, slice(d, 1:10)) -> m

slice(d, 11:20) %>%
  ggplot(aes(x, y)) +
  geom_point() +
  stat_function(fun = ~ predict(m, data.frame(x = .x)))
```

![](https://i.imgur.com/uRuOUKR.png)

<sup>Created on 2021-07-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by tmalsburg** [comment]
Thanks, @teunbrand. That's perhaps the best you can do currently, but perhaps not super obvious to most users and still quite a bit of typing given that we're asking a relatively straightforward thing.  Again in base-R we can just say `plot(d); abline(m)`.  However, this example suggests another possible solution:  `stat_function(fun=m)` since a `y~x` model *is* a way to represent a function.

**Comment by smouksassi** [comment]
would the above work with facets ?

**Comment by tmalsburg** [comment]
> No my proposed solution won't work with facets; or at least, will be repeated across facets, I think. 

Isn't a repetition of the model across facets what you would expect?  If not, what else would you expect to get?

> A counter-argument to the base R graphics point you're making, is that abline(m) only works with models that predict straight lines.

That's true but seems like an orthogonal point to me. If a ggplot solution works with more models than base-R that's simply strengthens the argument for such an solution, no?

**Comment by clauswilke** [comment]
A couple of thoughts from me: First, at a high level, while I agree the requested functionality is useful, I don't think we should tackle it in the base ggplot2 package. There's already more work than we can possibly do that has to happen at the base layer (themes, scales, coordinate systems, etc.) that anything that clearly can be handled elsewhere and by other developers should be handled there. That's the entire point of the extension system.

Now, some specific comments:

- `geom_smooth()` is the wrong place for this feature. The whole point of `geom_smooth()` is that it fits a model and then displays that. A geom that displays a pre-fitted model should have a different name.

- However, I'm not sure that using a geom is the correct approach at all. To me this is more like an annotation. I could be wrong though.

- Once somebody embarks down the path of implementing this, there will be hundreds of requests to support this model class and that model class and fitted draws and so on. This is a massive undertaking and definitely makes more sense in its own package.

**Comment by tmalsburg** [comment]
> `geom_smooth()` is the wrong place for this feature. The whole point of `geom_smooth()` is that it fits a model and then displays that. A geom that displays a pre-fitted model should have a different name.

Agreed. The description of `geom_smooth` is "*Aids the eye in seeing patterns in the presence of overplotting.*" in other words you expect the `geom_smooth` to fit a model on the data displayed in the plot, which is a different use case than what we're discussing here.

> However, I'm not sure that using a geom is the correct approach at all. To me this is more like an annotation. I could be wrong though.

The description of `geom_function` covers the present case nicely, though, since a pre-fitted model actually *is* a function.  Description of `geom_function`: "*Computes and draws a function as a continuous curve. This makes it easy to superimpose a function on top of an existing plot.*"  This is precisely what we'd like to achieve here.

> Once somebody embarks down the path of implementing this, there will be hundreds of requests to support this model class and that model class and fitted draws and so on.

Wouldn't it be enough to require that the model has a `predict` function and otherwise treat it as a black box?  This way, if someone wanted support for a new model class, they'd have to add that in the package of the model, not in ggplot.

**Comment by mjskay** [comment]
Having built stuff like this in {tidybayes} (though in a way that only really works for Bayesian models), I must agree with @clauswilke that a geom is probably the wrong place for this. One of the reasons is that as a solution it breaks down rapidly as soon as you have more complex models with additional predictors, since your model should be operating on columns from the original data and not columns as transformed by aesthetic mappings. I think you would effectively have to provide a reverse mapping from plot space back to data space to make it work.

I have found that the better solution is to construct a meaningful prediction grid in data space (not plot space). Then you can just use the same mappings from data space to plot space as needed for both the fit line geometry and the data geometry (usually only typing these once, in the call to `ggplot()`). It takes very few lines of code and is a quite flexible approach. You can see some examples [here](https://mjskay.github.io/tidybayes/articles/tidy-brms.html#fit-prediction-curves) or a description in video [here](https://www.youtube.com/watch?v=wbzfqh_3LyM#t=10m22s). It's inspired in part by {modelr}, which you could probably use to do something similar on non-Bayesian models (and which is built on top of the `predict()` function).
